lib/src/modules/task.cc

```diff
@@ -0,0 +1,332 @@
+        "required": ["filename", "uri", "sha256"]
+      },
+      "minItems": 1
+    },
+    "input": {
+      "type": "object"
+    },
+    "inputFormat": {
+      "type": "string"
+    }
+  },
+  "required": ["task", "files", "input"],
+  "additionalProperties": true
+}
+)" };
+
+static const std::string TASK_WRAPPER_EXECUTABLE { "task_wrapper" };
+static const int TASK_WRAPPER_FILE_ERROR_EC { 5 };
+
+// Hard-code interpreters on Windows. On non-Windows, we still rely on permissions and #!
+static const std::map<std::string, std::function<TaskCommand(std::string)>> BUILTIN_TASK_INTERPRETERS {
+#ifdef _WIN32
+    {".rb",  [](std::string filename) { return TaskCommand {
+        "ruby", { filename }
+    }; }},
+    {".pp",  [](std::string filename) { return TaskCommand {
+        "puppet", { "apply", filename }
+    }; }},
+    {".ps1", [](std::string filename) { return TaskCommand {
+        "powershell", { "-NoProfile", "-NonInteractive", "-NoLogo", "-ExecutionPolicy", "Bypass", "-File", filename }
+    }; }}
+#endif
+};
+
+Task::Task(const std::string& spool_dir) :
+    storage_ { spool_dir },
+    wrapper_executable_ { (Configuration::Instance().getExecPrefix() / TASK_WRAPPER_EXECUTABLE).string() }
+{
+    module_name = "task";
+    actions.push_back(TASK_RUN_ACTION);
+
+    PCPClient::Schema input_schema { TASK_RUN_ACTION, TASK_RUN_ACTION_INPUT_SCHEMA };
+    PCPClient::Schema output_schema { TASK_RUN_ACTION };
+
+    input_validator_.registerSchema(input_schema);
+    results_validator_.registerSchema(output_schema);
+}
+
+static void addParametersToEnvironment(const lth_jc::JsonContainer &input, std::map<std::string, std::string> &environment)
+{
+    for (auto& k : input.keys()) {
+        auto val = input.type(k) == lth_jc::String ? input.get<std::string>(k) : input.toString(k);
+        environment.emplace("PT_"+k, move(val));
+    }
+}
+
+static TaskCommand getTaskCommand(const fs::path &task_executable)
+{
+    auto builtin = BUILTIN_TASK_INTERPRETERS.find(task_executable.extension().string());
+    if (builtin != BUILTIN_TASK_INTERPRETERS.end()) {
+        return builtin->second(task_executable.string());
+    }
+
+    return TaskCommand { task_executable.string(), { } };
+}
+
+// Verify (this includes checking the SHA256 checksums) that *all* task files are present
+// in the task cache downloading them if necessary.
+// Return the full path of the cached version of the first file from the list (which
+// is assumed to be the task executable).
+static fs::path getCachedTaskFile(const std::vector<lth_jc::JsonContainer> &files) {
+    auto task_cache_dir = Configuration::Instance().getAgentConfiguration().task_cache_dir;
+    if (files.empty()) {
+        throw Module::ProcessingError {
+            lth_loc::format("at least one file must be specified for a task") };
+    }
+    auto file = files[0];
+    LOG_DEBUG("Verifying task file based on {1}", file.toString());
+
+    // TODO task file verification and downloading
```

Comment:
@ekinanp this is where I'd expect work from https://github.com/puppetlabs/pxp-agent/pull/617 to go.